server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java

```diff
@@ -476,73 +464,106 @@ protected MapperBuilderContext createChildContext(MapperBuilderContext mapperBui
     }
 
     public ObjectMapper merge(Mapper mergeWith, MergeReason reason, MapperBuilderContext parentBuilderContext) {
-        if ((mergeWith instanceof ObjectMapper) == false) {
-            MapperErrors.throwObjectMappingConflictError(mergeWith.name());
-        }
-        if (mergeWith instanceof NestedObjectMapper) {
-            // TODO stop NestedObjectMapper extending ObjectMapper?
-            MapperErrors.throwNestedMappingConflictError(mergeWith.name());
-        }
-        ObjectMapper mergeWithObject = (ObjectMapper) mergeWith;
-        ObjectMapper merged = clone();
-        merged.doMerge(mergeWithObject, reason, parentBuilderContext);
-        return merged;
+        var mergeResult = MergeResult.build(this, mergeWith, reason, parentBuilderContext);
+        return new ObjectMapper(
+            simpleName(),
+            fullPath,
+            mergeResult.enabled,
+            mergeResult.subObjects,
+            mergeResult.dynamic,
+            mergeResult.mappers
+        );
     }
 
-    protected void doMerge(final ObjectMapper mergeWith, MergeReason reason, MapperBuilderContext parentBuilderContext) {
-        if (mergeWith.dynamic != null) {
-            this.dynamic = mergeWith.dynamic;
-        }
+    protected record MergeResult(
+        Explicit<Boolean> enabled,
+        Explicit<Boolean> subObjects,
+        ObjectMapper.Dynamic dynamic,
+        Map<String, Mapper> mappers
+    ) {
 
-        if (mergeWith.enabled.explicit()) {
-            if (reason == MergeReason.INDEX_TEMPLATE) {
-                this.enabled = mergeWith.enabled;
-            } else if (isEnabled() != mergeWith.isEnabled()) {
-                throw new MapperException("the [enabled] parameter can't be updated for the object mapping [" + name() + "]");
+        public static MergeResult build(
+            ObjectMapper existing,
+            Mapper mergeWith,
+            MergeReason reason,
+            MapperBuilderContext parentBuilderContext
+        ) {
+            if ((mergeWith instanceof ObjectMapper) == false) {
+                MapperErrors.throwObjectMappingConflictError(mergeWith.name());
             }
-        }
-
-        if (mergeWith.subobjects.explicit()) {
-            if (reason == MergeReason.INDEX_TEMPLATE) {
-                this.subobjects = mergeWith.subobjects;
-            } else if (subobjects != mergeWith.subobjects) {
-                throw new MapperException("the [subobjects] parameter can't be updated for the object mapping [" + name() + "]");
+            if (existing instanceof NestedObjectMapper == false && mergeWith instanceof NestedObjectMapper) {
+                // TODO stop NestedObjectMapper extending ObjectMapper?
```

Comment:
Yes, this feels like a logical next step - NestedObjectMapper could contain an ObjectMapper rather than extending it, and that would remove a whole bunch of weird instanceof checks elsewhere too.